File: src/commands/count/undergrads.rs
use crate::prelude::*;

/// Count the number of registered undergraduate students.
#[poise::command(slash_command, guild_only)]
#[tracing::instrument(skip_all, err(Debug))]
pub async fn undergrads(context: BotContext<'_>) -> BotResult<()> {
    // Defer reply to allow time for counting students
    context.defer().await?;

    // Count the total and registered undergrads
    let count = sqlx::query!(
        "select count(registered_user_id) as registered, count(*) as total
        from users
        where kind = ?",
        VerifiedUserKind::Undergrad.to_string(),
    )
    .fetch_one(&context.data().database_pool)
    .await?;

    // Reply to the command
    context
        .send(|reply| {
            reply.embed(embeds::registration::count(
                "Undergraduates",
                count.registered as u64,
                count.total as u64,
            ))
        })
        .await?;

    Ok(())
}

File: norris/commands/count/undergrads.py
from discord import ApplicationContext
from sqlalchemy import func, select

from ...bot import Norris
from ...model import VerifiedUser, VerifiedUserKind


async def handle_count_undergrads(norris: Norris, context: ApplicationContext) -> None:
    """
    Handles the `count undergrads` slash command.
    """
    # Defer response to allow time for query
    await context.response.defer()

    # Count the total and registered undergrads
    async with norris.database_engine.begin() as connection:
        result = await connection.execute(
            select(func.count(VerifiedUser.registered_user_id), func.count())
            .where(VerifiedUser.kind == VerifiedUserKind.UNDERGRAD),
        )
        registered, total = result.one()

    from ...responses import embeds

    # Display results to user
    await context.followup.send(
        embed=embeds.registration.count("Undergraduates", registered, total),
    )
